whoops, better handle PLUS_ASSIGNMENT fullRoleCode += MODIFIER_DELIMITER + RoleModifier.readOnly

EP version `2.3.1`
NullAway version `0.4.2`

### What did you do?
`mvn clean compile`

```
An unhandled exception was thrown by the Error Prone static analysis plugin.
[ERROR] Please report this at https://github.com/google/error-prone/issues/new and include the following:
[ERROR] 
[ERROR] error-prone version: 2.3.1
[ERROR] BugPattern: NullAway
[ERROR] Stack Trace:
[ERROR] java.lang.RuntimeException: whoops, better handle PLUS_ASSIGNMENT fullRoleCode += MODIFIER_DELIMITER + RoleModifier.readOnly
[ERROR] at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:1676)
[ERROR] at com.uber.nullaway.NullAway.checkReturnExpression(NullAway.java:539)
[ERROR] at com.uber.nullaway.NullAway.matchReturn(NullAway.java:294)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:1012)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1548)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR] at com.sun.source.util.TreeScanner.visitIf(TreeScanner.java:419)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:755)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1427)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR] at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:527)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR] at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:895)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR] at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:587)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
[ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR] at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:602)
[ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:145)
[ERROR] at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
[ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
[ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:64)
[ERROR] at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
[ERROR] at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
[ERROR] at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
[ERROR] at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
[ERROR] at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1353)
[ERROR] at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:952)
[ERROR] at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
[ERROR] at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
[ERROR] at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
[ERROR] at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
[ERROR] at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:137)
[ERROR] at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:108)
[ERROR] at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:118)
[ERROR] at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne$CompilerInvoker.compile(JavacCompilerWithErrorProne.java:223)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne.performCompile(JavacCompilerWithErrorProne.java:91)
[ERROR] at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1075)
[ERROR] at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:168)
[ERROR] at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
[ERROR] at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
[ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
[ERROR] at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
[ERROR] at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
[ERROR] at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
[ERROR] at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
```

Oof thanks for the report. Should be an easy fix, will look soon.

On Tue, Apr 24, 2018, 07:53 Chris de Almeida <notifications@github.com>
wrote:

> EP version 2.3.1
> NullAway version 0.4.2
> What did you do?
>
> mvn clean compile
>
> An unhandled exception was thrown by the Error Prone static analysis plugin.
> [ERROR] Please report this at https://github.com/google/error-prone/issues/new and include the following:
> [ERROR]
> [ERROR] error-prone version: 2.3.1
> [ERROR] BugPattern: NullAway
> [ERROR] Stack Trace:
> [ERROR] java.lang.RuntimeException: whoops, better handle PLUS_ASSIGNMENT fullRoleCode += MODIFIER_DELIMITER + RoleModifier.readOnly
> [ERROR] at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:1676)
> [ERROR] at com.uber.nullaway.NullAway.checkReturnExpression(NullAway.java:539)
> [ERROR] at com.uber.nullaway.NullAway.matchReturn(NullAway.java:294)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:1012)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1548)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
> [ERROR] at com.sun.source.util.TreeScanner.visitIf(TreeScanner.java:419)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:755)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1427)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
> [ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
> [ERROR] at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:527)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
> [ERROR] at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:895)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
> [ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
> [ERROR] at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:587)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
> [ERROR] at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
> [ERROR] at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
> [ERROR] at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:602)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:145)
> [ERROR] at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
> [ERROR] at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
> [ERROR] at com.google.errorprone.scanner.Scanner.scan(Scanner.java:64)
> [ERROR] at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
> [ERROR] at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
> [ERROR] at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
> [ERROR] at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
> [ERROR] at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1353)
> [ERROR] at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:952)
> [ERROR] at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
> [ERROR] at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
> [ERROR] at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
> [ERROR] at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
> [ERROR] at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:137)
> [ERROR] at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:108)
> [ERROR] at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:118)
> [ERROR] at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne$CompilerInvoker.compile(JavacCompilerWithErrorProne.java:223)
> [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> [ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> [ERROR] at java.lang.reflect.Method.invoke(Method.java:498)
> [ERROR] at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne.performCompile(JavacCompilerWithErrorProne.java:91)
> [ERROR] at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1075)
> [ERROR] at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:168)
> [ERROR] at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
> [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
> [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
> [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
> [ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
> [ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
> [ERROR] at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
> [ERROR] at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
> [ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
> [ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
> [ERROR] at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
> [ERROR] at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
> [ERROR] at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
> [ERROR] at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
> [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> [ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> [ERROR] at java.lang.reflect.Method.invoke(Method.java:498)
> [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
> [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
> [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
> [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uber/NullAway/issues/156>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALyUZPezB7BWWyF0k4M0Pk4dyTGQMlWks5trzx7gaJpZM4Th1zA>
> .
>

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -56,6 +56,7 @@ import com.sun.source.tree.AssignmentTree;
 import com.sun.source.tree.BinaryTree;
 import com.sun.source.tree.BlockTree;
 import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompoundAssignmentTree;
 import com.sun.source.tree.ConditionalExpressionTree;
 import com.sun.source.tree.ExpressionStatementTree;
 import com.sun.source.tree.ExpressionTree;
@@ -162,7 +163,8 @@ public class NullAway extends BugChecker
         BugChecker.ForLoopTreeMatcher,
         BugChecker.LambdaExpressionTreeMatcher,
         BugChecker.IdentifierTreeMatcher,
-        BugChecker.MemberReferenceTreeMatcher {
+        BugChecker.MemberReferenceTreeMatcher,
+        BugChecker.CompoundAssignmentTreeMatcher {
 
   private static final String INITIALIZATION_CHECK_NAME = "NullAway.Init";
 
@@ -372,6 +374,20 @@ public class NullAway extends BugChecker
       String message = "assigning @Nullable expression to @NonNull field";
       return createErrorDescriptionForNullAssignment(
           MessageTypes.ASSIGN_FIELD_NULLABLE, tree, message, expression, state.getPath());
+    }
+    return Description.NO_MATCH;
+  }
+
+  @Override
+  public Description matchCompoundAssignment(CompoundAssignmentTree tree, VisitorState state) {
+    if (!matchWithinClass) {
+      return Description.NO_MATCH;
+    }
+    Type lhsType = ASTHelpers.getType(tree.getVariable());
+    Type stringType = state.getTypeFromString("java.lang.String");
+    if (lhsType != null && !lhsType.equals(stringType)) {
+      // both LHS and RHS could get unboxed
+      return doUnboxingCheck(state, tree.getVariable(), tree.getExpression());
     }
     return Description.NO_MATCH;
   }
@@ -1626,6 +1642,18 @@ public class NullAway extends BugChecker
         // Lambdas may return null, but the lambda literal itself should not be null
       case MEMBER_REFERENCE:
         // These cannot be null; the compiler would catch it
+      case MULTIPLY_ASSIGNMENT:
+      case DIVIDE_ASSIGNMENT:
+      case REMAINDER_ASSIGNMENT:
+      case PLUS_ASSIGNMENT:
+      case MINUS_ASSIGNMENT:
+      case LEFT_SHIFT_ASSIGNMENT:
+      case RIGHT_SHIFT_ASSIGNMENT:
+      case UNSIGNED_RIGHT_SHIFT_ASSIGNMENT:
+      case AND_ASSIGNMENT:
+      case XOR_ASSIGNMENT:
+      case OR_ASSIGNMENT:
+        // result of compound assignment cannot be null
       case PLUS:
         // rest are for auto-boxing
       case MINUS:

--- a/nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java
+++ b/nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java
@@ -465,4 +465,26 @@ public class NullAwayTest {
             "}")
         .doTest();
   }
+
+  @Test
+  public void compoundAssignment() {
+    compilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "class Test {",
+            "  static void assignments() {",
+            "    String x = null; x += \"hello\";",
+            "    // BUG: Diagnostic contains: unboxing of a @Nullable value",
+            "    Integer y = null; y += 3;",
+            "    // BUG: Diagnostic contains: unboxing of a @Nullable value",
+            "    boolean b = false; Boolean c = null; b |= c;",
+            "  }",
+            "  static Integer returnCompound() {",
+            "    Integer z = 7;",
+            "    return (z += 10);",
+            "  }",
+            "}")
+        .doTest();
+  }
 }

